solard/solard/tcp_server.py

```diff
@@ -0,0 +1,291 @@
+#    under the License.
+
+from gevent import monkey
+monkey.patch_all()
+
+
+from gevent.server import StreamServer
+import socket
+
+import msgpack
+import struct
+import errno
+import sys
+import traceback
+import pwd
+import os
+
+from types import GeneratorType
+from solard.logger import logger
+from solard.core import SolardContext, SolardIface
+
+
+SERVER_BUFF = 4096
+HDR = "<I"
+HDR_SIZE = struct.calcsize(HDR)
+
+INT_DEFAULT_REPLY_TYPE = 0
+INT_GENERATOR_REPLY_TYPE = 1
+
+
+class SolardTCPException(Exception):
+    pass
+
+
+class ReadFailure(SolardTCPException):
+    pass
+
+
+class SolardTCPHandler(object):
+
+    def __init__(self, sock, address):
+        self.sock = sock
+        self.address = address
+        self.ctx = SolardContext()
+        self.auth = None
+        self._wrote = False
+        self.forked = False
+
+    def _read(self):
+        # TODO: client closed connection
+        try:
+            size = struct.unpack(HDR, self.sock.recv(HDR_SIZE))[0]
+        except:
+            raise ReadFailure("Can't read header data")
+        d = []
+        while True:
+            b = min(size, SERVER_BUFF)
+            curr = self.sock.recv(b)
+            if not curr:
+                raise ReadFailure("No data")
+            d.append(curr)
+            size -= len(curr)
+            assert size >= 0
+            if not size:
+                break
+        self._wrote = False
+        try:
+            return msgpack.unpackb(''.join(d))
+        except:
+            print '--------------------'
+            print repr(d)
+            print '--------------------'
+            raise
+
+    def _read_stream(self, size=None):
+        if size is None:
+            try:
+                size = struct.unpack(HDR, self.sock.recv(HDR_SIZE))[0]
+            except:
+                raise ReadFailure("Can't read header data")
+        while True:
+            b = min(size, SERVER_BUFF)
+            curr = self.sock.recv(b)
+            if not curr:
+                if size > 0:
+                    raise ReadFailure("Expected more data")
+            size -= len(curr)
+            assert size >= 0
+            yield curr
+            if not size:
+                break
+
+    def _write(self, **kwargs):
+        assert self._wrote is False
+        _data = msgpack.packb(kwargs)
+        size = len(_data)
+        hdr = struct.pack(HDR, size)
+        self.sock.sendall(hdr + _data)
+        self._wrote = True
+
+    def _write_ok(self, res):
+        # logger.debug("Ok sent")
+        data = {'st': 2, 'res': res}
+        self._write(**data)
+
+    def _write_ok_gen(self, res):
+        data = {'st': 20, 'res': res}
+        self._write(**data)
+
+    # def _write_ok_stream(self, res):
+    #     data = {'st': 30, 'res': res}
+    #     self._write(**data)
+
+    # def _write_stream_data(self, data):
+    #     self.sock.sendall(data)
+
+    def _write_gen_end(self):
+        data = {'st': 21, 'res': None}
```

Comment:
this codes - 'st' - it will be like our convention? if so consts or at least message would be nice